### PR TITLE
Speedup improvements

### PR DESCRIPTION
**Pull request by cremebrule**

### CODE REVIEW

File: omnigibson/object_states/object_state_base.py [hunk 1]
```diff
@@ -67,23 +67,37 @@ def reset(self):
...
+    def requires_update(cls):
+        """
+        Returns:
+            bool: True if this state requires an update or not every sim step
+        """
+        # The __repr__ call tells us which actual function is being called
+        # Any subclass that actually implements _update() will cause this to return its own class_name._update
+        # Otherwise, __repr__ will return BaseObjectState._update
+        # So we use this as a heuristic to check whether or not a subclass has overridden the _update method
+        return "BaseObjectState" not in cls._update.__repr__()
```
Comment (ChengshuLi) [inline omnigibson/object_states/object_state_base.py:93]: hmm this seems kind of hacky... can we find an alternative?

File: omnigibson/object_states/object_state_base.py [hunk 2]
```diff
@@ -67,23 +67,37 @@ def reset(self):
...
         """
         Returns:
-            OrdereDict: Dictionary mapping specific argument combinations from @self.get_value() to cached values and
+            dict: Dictionary mapping specific argument combinations from @self.get_value() to cached values and
                 information stored for that specific combination
         """
         return self._cache
 
+    @classproperty
+    def requires_update(cls):
```
Comment (ChengshuLi) [inline omnigibson/object_states/object_state_base.py:84]: where is this called?

File: omnigibson/object_states/object_state_base.py [hunk 3]
```diff
@@ -47,6 +46,7 @@ def __init__(self, obj):
         self._cache = None
         self._changed = None
         self._simulator = None
+        self._last_t_updated = -1               # Last timestep when this state was updated
```
Comment (ChengshuLi) [inline omnigibson/object_states/object_state_base.py:49]: sorry I am not sure I understand why we need this

File: omnigibson/object_states/particle_modifier.py [hunk 4]
```diff
@@ -550,7 +570,7 @@ def _modify_particles(self, system):
             if system.n_particles > 0:
                 # Iterate over all particles and remove any that are within the relaxed AABB of the remover volume
                 particle_names = list(system.particles.keys())
-                particle_positions = np.array([particle.get_position() for particle in system.particles.values()])
+                particle_positions = np.array([system.get_particle_position_orientation(name=name) for name in system.particles.keys()])
```
Comment (ChengshuLi) [inline omnigibson/object_states/particle_modifier.py:573]: should be `system.get_particle_position_orientations(name=name)[0]`?

File: omnigibson/prims/entity_prim.py [hunk 5]
```diff
@@ -83,6 +81,9 @@ def _initialize(self):
         # Update joint information
         self.update_joints()
 
+        # Update the spawn pose of this prim now that it's been initialized properly
+        self.update_spawn_position_orientation()
```
Comment (ChengshuLi) [inline omnigibson/prims/entity_prim.py:85]: why is this needed?

File: omnigibson/simulator.py [hunk 6]
```diff
@@ -366,7 +367,7 @@ def _non_physics_step(self):
 
             # Step the object states in global topological order (if the scene exists).
             if self.scene is not None:
-                for state_type in self.object_state_types:
+                for state_type in self.object_state_types_requiring_update:
```
Comment (ChengshuLi) [inline omnigibson/simulator.py:370]: does this actually bring performance improvements?

File: omnigibson/simulator.py [hunk 7]
```diff
@@ -478,28 +479,40 @@ def reset_scene(self):
...
+
+            # Update all object handles and spawn poses
+            if self.scene is not None and self.scene.initialized:
+                for obj in self.scene.objects:
+                    # Only need to update if object is already initialized as well
+                    if obj.initialized:
+                        obj.update_handles()
+                        # Only need to update spawn pose if we were stopped
+                        if was_stopped:
+                            obj.update_spawn_position_orientation()
```
Comment (ChengshuLi) [inline omnigibson/simulator.py:498]: why is this needed?

File: omnigibson/systems/macro_particle_system.py [hunk 8]
```diff
@@ -643,13 +645,24 @@ def generate_group_particles(
...
                 position -= normal * base_to_center
 
+            if gm.ENABLE_FLATCACHE:
+                # If flatcache is enabled, we need to modify the position we're setting because we're directly writing
+                # to the USD, which is NOT being updated when flatcache is enabled!
+                # So, we need to set the particle w.r.t. the link's initial (i.e.: spawn) pose, NOT the current pose!
+                # We will calculate the transform from the link's current pose to its initial pose, and then apply this
+                # transform to the desired global pose
+                tf = T.pose_inv(T.pose2mat(link.get_position_orientation())) @ T.pose2mat(link.spawn_position_orientation)
+                position, orientation = T.mat2pose(tf @ T.pose2mat((position, orientation)))
```
Comment (ChengshuLi) [inline omnigibson/systems/macro_particle_system.py:664]: sorry not sure what this means, maybe we can discuss in person?

File: omnigibson/utils/usd_utils.py [hunk 9]
```diff
@@ -248,12 +249,93 @@ class BoundingBoxAPI:
...
+            # If the simulation is not stopped and we're a rigid prim, then we need to manually compute the real-time
+            # transform from the prim's initial pose to the current pose
+            if isinstance(prim, RigidPrim) and not og.sim.is_stopped():
+                val = np.stack(
+                    [arr.flatten() for arr in np.meshgrid(*(np.linspace(lo, hi, 2) for lo, hi in zip(*val)))] + [
+                        np.ones(8, )])
+                # tf = T.pose2mat(prim.get_position_orientation()) @ T.pose2mat(prim.spawn_position_orientation)
+                tf = T.pose_inv(T.pose2mat(prim.spawn_position_orientation)) @ T.pose2mat(prim.get_position_orientation())
+                val = (tf @ val)[:-1, :]
+                val = np.min(val, axis=1), np.max(val, axis=1)
```
Comment (ChengshuLi) [inline omnigibson/utils/usd_utils.py:322]: sorry I am not sure I follow what is happening here

File: omnigibson/envs/env_base.py [hunk 10]
```diff
@@ -169,7 +168,10 @@ def _load_task(self):
         assert og.sim.is_stopped(), "sim should be stopped when load_task starts"
         og.sim.play()
 
-        # Load task. Should load additinal task-relevant objects and configure the scene into its default initial state
+        # Load the state from the loaded scene
+        og.sim.reset_scene()
```
[the remark this answers is not in the document]
Comment (ChengshuLi) [reply omnigibson/envs/env_base.py:172]: conclusion: yes it's needed, maybe add more comments.

File: omnigibson/object_states/object_state_base.py (hunk 1 above)
Comment (ChengshuLi) [reply omnigibson/object_states/object_state_base.py:93]: UpdatingState?

File: omnigibson/envs/env_base.py (hunk 10 above)
[the remark this answers is not in the document]
Comment (cremebrule) [reply omnigibson/envs/env_base.py:172]: done!

File: omnigibson/simulator.py (hunk 6 above)
Comment (ChengshuLi) [reply omnigibson/simulator.py:370]: yes!

File: omnigibson/object_states/object_state_base.py [hunk 11]
```diff
@@ -109,16 +123,17 @@ def update(self):
         Updates the object state, possibly clearing internal cached information
         """
         assert self._initialized, "Cannot update uninitialized state."
-        # Clear all the changed values
-        self._changed = OrderedDict()
+        # Clear all the changed values and update the last t value
+        self._changed = dict()
```
Comment (ChengshuLi) [inline omnigibson/object_states/object_state_base.py:127]: this line is not needed

File: omnigibson/object_states/object_state_base.py (hunk 3 above)
Comment (cremebrule) [reply omnigibson/object_states/object_state_base.py:49]: resolved?

File: omnigibson/object_states/particle_modifier.py (hunk 4 above)
Comment (cremebrule) [reply omnigibson/object_states/particle_modifier.py:565]: done!

File: omnigibson/simulator.py (hunk 7 above)
Comment (cremebrule) [reply omnigibson/simulator.py]: no longer needed :D

File: omnigibson/utils/usd_utils.py (hunk 9 above)
Comment (cremebrule) [reply omnigibson/utils/usd_utils.py]: resolved

File: omnigibson/systems/macro_particle_system.py (hunk 8 above)
Comment (cremebrule) [reply omnigibson/systems/macro_particle_system.py]: resolved!

File: omnigibson/prims/entity_prim.py (hunk 5 above)
Comment (cremebrule) [reply omnigibson/prims/entity_prim.py]: not needed!

File: omnigibson/object_states/object_state_base.py (hunk 11 above)
Comment (cremebrule) [reply omnigibson/object_states/object_state_base.py:127]: done!

File: omnigibson/object_states/object_state_base.py (hunk 1 above)
Comment (cremebrule) [reply omnigibson/object_states/object_state_base.py:93]: done!

File: omnigibson/scenes/scene_base.py [hunk 12]
```diff
@@ -602,6 +600,11 @@ def _load_state(self, state):
         # Default state for the scene is from the registry alone
         self._registry.load_state(state=state, serialized=False)
 
+        # We additionally clear caches for all object states
+        for obj in self.objects:
+            if isinstance(obj, StatefulObject):
+                obj.clear_states_cache()
```
Comment (ChengshuLi) [inline omnigibson/scenes/scene_base.py:606]: I think in `obj._load_state`, we should do `clear_states_cache`?

File: omnigibson/simulator.py [hunk 13]
```diff
@@ -478,28 +485,41 @@ def reset_scene(self):
...
+                        obj.update_handles()
+
+            # Check to see if any objects should be initialized
+            if len(self._objects_to_initialize) > 0:
+                for obj in self._objects_to_initialize:
+                    obj.initialize()
+                self._objects_to_initialize = []
+                # Also update the scene registry
+                # TODO: A better place to put this perhaps?
+                self._scene.object_registry.update(keys="root_handle")
```
Comment (ChengshuLi) [inline omnigibson/simulator.py:510]: sorry why is this needed?

File: omnigibson/simulator.py [hunk 14]
```diff
@@ -478,28 +485,41 @@ def reset_scene(self):
...
-        if len(self._objects_to_initialize) > 0:
-            for obj in self._objects_to_initialize:
-                obj.initialize()
-            self._objects_to_initialize = []
-            # Also update the scene registry
-            # TODO: A better place to put this perhaps?
-            self._scene.object_registry.update(keys="root_handle")
+        # If we're using flatcache, we also need to reset its API
+        if gm.ENABLE_FLATCACHE:
+            FlatcacheAPI.reset()
```
Comment (ChengshuLi) [inline omnigibson/simulator.py:522]: why does reset do?

File: omnigibson/utils/usd_utils.py [hunk 15]
```diff
@@ -343,6 +393,108 @@ def aabb_contains_point(cls, point, container):
...
+        Args:
+            prim (EntityPrim): prim whose owned links and joints should have their raw local states updated to match the
+                "true" values found from the dynamic control interface
+        """
+        # Make sure flatcache is enabled -- this should NEVER be called otherwise!!
+        assert gm.ENABLE_FLATCACHE, "Syncing raw object transforms should only occur if flatcache is being used!"
+
+        # We're somewhat abusing low-level dynamic control - physx - usd integration, but we (supposedly) know
+        # what we're doing so we suppress logging so we don't see any error messages :D
+        with suppress_logging(["omni.physx.plugin"]):
```
Comment (ChengshuLi) [inline omnigibson/utils/usd_utils.py:423]: omg this is great!

**Review by ChengshuLi — APPROVED** [review]

**Review by ChengshuLi — APPROVED** [review]

**Comment by ChengshuLi** [comment]
LGTM! Feel free to merge!
